src/conn/conn_compact.c

```diff
@@ -0,0 +1,202 @@
+ *     Check to decide if the compact server should continue running.
+ */
+static bool
+__compact_server_run_chk(WT_SESSION_IMPL *session)
+{
+    return (FLD_ISSET(S2C(session)->server_flags, WT_CONN_SERVER_COMPACT));
+}
+
+/*
+ * __compact_server --
+ *     The compact server thread.
+ */
+static WT_THREAD_RET
+__compact_server(void *arg)
+{
+    WT_CONNECTION_IMPL *conn;
+    WT_CURSOR *cursor;
+    WT_DECL_RET;
+    WT_SESSION *wt_session;
+    WT_SESSION_IMPL *session;
+    int exact;
+    const char *config, *key;
+    bool full_iteration;
+
+    session = arg;
+    conn = S2C(session);
+    wt_session = (WT_SESSION *)session;
+    cursor = NULL;
+    config = NULL;
+    key = NULL;
+    exact = 0;
+    full_iteration = false;
+
+    for (;;) {
+
+        /* When the entire metadata file has been parsed, take a break. */
+        if (full_iteration) {
+            full_iteration = false;
+            /*
+             * TODO: We may have processed just a few files, might as well sleep to give the system
+             * a break.
+             */
+            __wt_sleep(10, 0);
+        }
+
+        /* Only an explicit call to the compact API should signal the thread to wake up. */
+        if (!conn->background_compact.running) {
+            /*
+             * TODO: Depending on the previous state, we may not want to clear out the last key
+             * used. This could be useful if the server was paused to be resumed later.
+             */
+            key = NULL;
+            __wt_cond_wait(session, conn->background_compact.cond, 0, NULL);
+        }
+
+        /*
+         * Now the server has been requested to run, check if it is supposed to, we could be
+         * quitting or reconfiguring the connection too.
+         */
+        if (!__compact_server_run_chk(session))
+            break;
+
+        /* Open a metadata cursor. */
+        WT_ERR(__wt_metadata_cursor(session, &cursor));
+
+        /*
+         * If we are in the middle of an iteration, position the cursor on the previously selected
+         * key.
+         */
+        if (key != NULL) {
+            cursor->set_key(cursor, key);
+            WT_ERR(cursor->search_near(cursor, &exact));
+        }
+
+        /*
+         * If no key has been set yet, move to the first record. If the previously used key could
+         * not be found, make sure not to go backwards.
+         */
+        if (key == NULL || exact <= 0)
+            WT_ERR_NOTFOUND_OK(cursor->next(cursor), true);
+
+        /* Find a table to compact. */
+        while (ret == 0) {
+            WT_ERR(cursor->get_key(cursor, &key));
+            if (WT_PREFIX_MATCH(key, "file:") && !WT_STREQ(key, WT_HS_URI))
+                /* FIXME-WT-11343: check if the table is supposed to be compacted. */
+                break;
+            WT_ERR_NOTFOUND_OK(cursor->next(cursor), true);
+        }
+
+        /* We may have reached the end of the file. */
+        if (ret == WT_NOTFOUND) {
```

Comment:
We'll be done as soon as the key lands on something with a prefix that is after `file:`. So we could quit then, rather than iterating to the end of the metadata file.

Reply:
That's right! Changes done in ed8fc62b692e23a3d5817136458e718884a6f20f.